version: "3.3"

services:
  traefik:
    image: "traefik:v2.10"
    restart: always
    logging:
      driver: "json-file"
      options:
        max-size: "10m"
        max-file: "10"
    user: ${USER_UID:?err}:${USER_GID:?err}
    ports:
      - "443:443"
      - "8080:8080"
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
      - "./letsencrypt:/letsencrypt"
    command:
      - "--api.insecure=true"
      - "--providers.docker=true"
      - "--providers.docker.exposedbydefault=false"
      #- "--entrypoints.web.address=:80"
      #- "--entrypoints.web.http.redirections.entrypoint.to=websecure"
      #- "--entrypoints.web.http.redirections.entrypoint.scheme=https"
      #- "--entrypoints.web.http.redirections.entrypoint.permanent=true"
      - "--entrypoints.websecure.address=:443"
      - "--certificatesresolvers.osom_resolver.acme.tlschallenge=true"
      - "--certificatesresolvers.osom_resolver.acme.email=${ACME_EMAIL:?err}"
      - "--certificatesresolvers.osom_resolver.acme.storage=/letsencrypt/acme.json"
      #-"--log.level=DEBUG"

  master:
    image: "ghcr.io/osom8979/osom-api:latest"
    restart: always
    depends_on:
      - traefik
    logging:
      driver: "json-file"
      options:
        max-size: "10m"
        max-file: "10"
    networks:
      - osom-net
    user: ${USER_UID:?err}:${USER_GID:?err}
    deploy:
      mode: replicated
      replicas: 2
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.osom_master.rule=Host(`${FRONTEND_HOST:?err}`)"
      - "traefik.http.routers.osom_master.entrypoints=websecure"
      - "traefik.http.routers.osom_master.tls.certresolver=osom_resolver"
      - "traefik.http.services.osom_master.loadbalancer.server.port=${HTTP_PORT:?err}"
    environment:
      - "TZ=Asia/Seoul"
      - "COLORED_LOGGING=True"
      - "USE_UVLOOP=${USE_UVLOOP:?err}"
      - "SEVERITY=${SEVERITY:?err}"
      - "DEBUG=${DEBUG:?err}"
      - "VERBOSE=${VERBOSE:?err}"
      - "HTTP_HOST=${HTTP_HOST:?err}"
      - "HTTP_PORT=${HTTP_PORT:?err}"
      - "HTTP_TIMEOUT=${HTTP_TIMEOUT:?err}"
      - "REDIS_HOST=${REDIS_HOST:?err}"
      - "REDIS_PORT=${REDIS_PORT:?err}"
      - "REDIS_DATABASE=${REDIS_DATABASE:?err}"
      - "REDIS_PASSWORD=${REDIS_PASSWORD:?err}"
      - "REDIS_USE_TLS=False"
      - "S3_ENDPOINT=${S3_ENDPOINT:?err}"
      - "S3_ACCESS=${S3_ACCESS:?err}"
      - "S3_SECRET=${S3_SECRET:?err}"
      - "S3_REGION=${S3_REGION:?err}"
      - "S3_BUCKET=${S3_BUCKET:?err}"
      - "SUPABASE_URL=${SUPABASE_URL:?err}"
      - "SUPABASE_KEY=${SUPABASE_KEY:?err}"
      - "HEALTHCHECK_TIMEOUT=${HEALTHCHECK_TIMEOUT:?err}"
      - "HEALTHCHECK_URI=http://localhost:${HTTP_PORT:?err}/health"
    healthcheck:
      test: ["CMD", "python3", "-m", "osom-api", "health"]
      interval: 10s
      timeout: 5s
      retries: 5
    command: master

networks:
  osom-net:
    external: true
